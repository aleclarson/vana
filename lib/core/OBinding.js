"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var shared_1 = require("../shared");
var Observable_1 = require("./Observable");
/**
 * An observable binding to some other stream-like mechanism.
 *
 * The binding is lazy, which means no value exists until observed.
 *
 * The current value is always `undefined` when no observers exist.
 */
var OBinding = /** @class */ (function (_super) {
    tslib_1.__extends(OBinding, _super);
    function OBinding(driver) {
        var _this = _super.call(this, undefined) || this;
        _this.driver = driver;
        return _this;
    }
    OBinding.prototype.activate = function () {
        var _this = this;
        var result = this.driver(function (newValue) { return _this.set(newValue, true); });
        if (result) {
            this._binding = shared_1.isFunction(result) ? { dispose: result } : result;
        }
    };
    OBinding.prototype.deactivate = function () {
        this._value = undefined;
        if (this._binding) {
            this._binding.dispose();
            this._binding = undefined;
        }
    };
    return OBinding;
}(Observable_1.Observable));
exports.OBinding = OBinding;
