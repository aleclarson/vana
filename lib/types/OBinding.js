"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Observable_1 = require("./Observable");
/**
 * An observable binding to some other stream-like mechanism.
 *
 * The binding is lazy, which means no value exists until observed.
 *
 * The current value is always `undefined` when no observers exist.
 */
var OBinding = /** @class */ (function (_super) {
    tslib_1.__extends(OBinding, _super);
    function OBinding(driver) {
        var _this = _super.call(this, undefined) || this;
        _this.driver = driver;
        return _this;
    }
    OBinding.prototype.activate = function () {
        var result = this.driver(this.set.bind(this));
        if (result) {
            this._binding = typeof result == 'function' ? { dispose: result } : result;
        }
    };
    OBinding.prototype.deactivate = function () {
        this._value = undefined;
        if (this._binding) {
            this._binding.dispose();
            this._binding = undefined;
        }
    };
    return OBinding;
}(Observable_1.Observable));
exports.OBinding = OBinding;
