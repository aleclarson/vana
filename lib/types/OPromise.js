"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/* tslint:disable:unified-signatures */
var common_1 = require("../common");
var commit_1 = require("../funcs/commit");
var freeze_1 = require("../funcs/freeze");
var symbols_1 = require("../symbols");
var Observable_1 = require("./Observable");
var freezeState = freeze_1.freeze;
var PENDING = freezeState({ resolved: false });
var OPromise = /** @class */ (function (_super) {
    tslib_1.__extends(OPromise, _super);
    function OPromise(resolver) {
        var _this = _super.call(this, function () { return _this._state; }) || this;
        _this._state = PENDING;
        if (resolver) {
            try {
                resolver(_this.resolve.bind(_this), _this.reject.bind(_this));
            }
            catch (e) {
                _this.reject(e);
            }
        }
        return _this;
    }
    OPromise.prototype.resolve = function (result) {
        if (this._state !== PENDING)
            return;
        if (common_1.isThenable(result)) {
            // Create our own pending state to prevent other `resolve` and `reject` calls.
            this._state = freezeState({ resolved: false });
            result
                .then(this._fulfill.bind(this), this._reject.bind(this))
                .then(common_1.noop, console.error); // tslint:disable-line
        }
        else {
            this._fulfill(result);
        }
    };
    OPromise.prototype.reject = function (error) {
        if (this._state === PENDING) {
            this._reject(error);
        }
    };
    /** @internal */
    OPromise.prototype['_isCurrent'] = function (value) {
        // OPromise instances are never bound to more than one Promise.
        return value && value[symbols_1.$O] === this;
    };
    /** Create a promise and resolve it immediately */
    OPromise.resolve = function (result) {
        var p = new OPromise();
        p.resolve(result);
        return p;
    };
    OPromise.prototype._fulfill = function (result) {
        this._state = freezeState({ resolved: true, result: result, error: undefined });
        commit_1.commit(this, PENDING, this._state);
    };
    OPromise.prototype._reject = function (error) {
        this._state = freezeState({ resolved: true, result: undefined, error: error });
        commit_1.commit(this, PENDING, this._state);
    };
    return OPromise;
}(Observable_1.Observable));
exports.OPromise = OPromise;
/** @internal */
function bindPromise(promise) {
    var observable = Observable_1.getObservable(promise);
    if (!observable) {
        observable = OPromise.resolve(promise);
        common_1.definePrivate(promise, symbols_1.$O, observable);
    }
    return observable;
}
exports.bindPromise = bindPromise;
