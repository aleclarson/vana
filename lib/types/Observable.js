"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var common_1 = require("../common");
var debug_1 = require("../debug");
var commit_1 = require("../funcs/commit");
var symbols_1 = require("../symbols");
var nextId = 1;
/** An observable value */
var Observable = /** @class */ (function () {
    function Observable(source) {
        this['_observers'] = undefined;
        this.id = nextId++;
        this._rebind(source);
    }
    /** Try coercing the given value into an Observable instance. */
    Observable.from = function (value) {
        return value instanceof Observable ? value : value ? value[symbols_1.$O] : undefined;
    };
    /** The current value. */
    Observable.prototype.get = function () {
        return this._get ? this._get() : this._value;
    };
    /**
     * When the source of truth is a getter, this method only notifies observers.
     * Otherwise, this method also updates the current value.
     */
    Observable.prototype.set = function (newValue) {
        var oldValue = this.get();
        if (newValue !== oldValue) {
            if (!this._get) {
                this._value = newValue;
            }
            commit_1.commit(this, oldValue, newValue);
        }
    };
    /** Listen for changes. */
    Observable.prototype.tap = function (onUpdate) {
        var _this = this;
        var observer = {
            onUpdate: onUpdate,
            dispose: function () { return _this._removeObserver(observer); },
            _onChange: function (change) { return change.prop !== null || onUpdate(change.newValue); },
        };
        this._addObserver(observer);
        return observer;
    };
    /** Remove all observers and perform subclass cleanup (if necessary) */
    Observable.prototype.dispose = function () {
        if (this._observers) {
            this._observers = undefined;
            this.deactivate();
        }
    };
    /** @internal Notify observers of a change */
    Observable.prototype['_onChange'] = function (change) {
        var observers = this._observers;
        if (debug_1.getDebug(this)) {
            // tslint:disable-next-line
            console.debug('%s(%s) changed: (observers: %O, change: %O)', this.constructor.name, this.id, observers ? observers.length : 0, common_1.shallowCopy(change) // TODO: Deep copy instead
            );
        }
        if (observers) {
            for (var _i = 0, _a = observers.slice(0); _i < _a.length; _i++) {
                var observer = _a[_i];
                observer._onChange(change);
            }
        }
    };
    /** @internal Returns false when the given value is outdated */
    Observable.prototype['_isCurrent'] = function (value) {
        return value === this.get();
    };
    /** @internal Swap out the source of truth for this observable. */
    Observable.prototype['_rebind'] = function (source) {
        this._get = typeof source == 'function' ? source : undefined;
        this._value = this._get ? undefined : source;
    };
    /** @internal */
    Observable.prototype['_addObserver'] = function (observer, prepend) {
        if (this._observers) {
            this._observers[prepend ? 'unshift' : 'push'](observer);
        }
        else {
            this.activate();
            this._observers = [observer];
        }
        return observer;
    };
    /** @internal */
    Observable.prototype['_removeObserver'] = function (observer) {
        var observers = this._observers;
        if (observers) {
            var index = observers.indexOf(observer);
            if (index < 0)
                return;
            if (observers.length > 1) {
                observers.splice(index, 1);
            }
            else {
                this._observers = undefined;
                this.deactivate();
            }
        }
    };
    /** Called when an Observable goes from 0 observers to 1+ */
    Observable.prototype.activate = function () { };
    /** Called when an Observable goes from 1+ observers to 0 */
    Observable.prototype.deactivate = function () { };
    return Observable;
}());
exports.Observable = Observable;
/** Returns true if the given value has been passed to `o()` and is not stale. */
function isObservable(value) {
    if (value) {
        var observable = value[symbols_1.$O];
        if (observable) {
            return observable._isCurrent(value);
        }
    }
    return false;
}
exports.isObservable = isObservable;
/** Throws an error when the given object is an old revision */
function getObservable(value) {
    var observable = value[symbols_1.$O];
    if (observable) {
        if (!observable._isCurrent(value)) {
            throw Error('Outdated values cannot be observed');
        }
        return observable;
    }
}
exports.getObservable = getObservable;
/** @internal */
function getObservers(value) {
    var observable = Observable.from(value);
    if (observable) {
        return observable._observers;
    }
    throw Error('Not observable');
}
exports.getObservers = getObservers;
