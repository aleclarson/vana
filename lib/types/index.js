"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Observable_1 = require("./Observable");
exports.Observable = Observable_1.Observable;
var OBinding_1 = require("./OBinding");
exports.OBinding = OBinding_1.OBinding;
var OPath_1 = require("./OPath");
exports.OPath = OPath_1.OPath;
var OPromise_1 = require("./OPromise");
exports.OPromise = OPromise_1.OPromise;
var OProps_1 = require("./OProps");
exports.OProps = OProps_1.OProps;
var OTuple_1 = require("./OTuple");
exports.OTuple = OTuple_1.OTuple;
