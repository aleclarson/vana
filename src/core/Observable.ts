import { getDebug } from '../debug'
import { $O, IDisposable, isFunction, shallowCopy } from '../shared'
import { Change, IChangeTarget } from './Change'
import { commit } from './commit'

/**
 * Observers are disposable "change targets".
 *
 * Use `IChangeTarget<T>` if your observer is not disposable.
 */
export interface IObserver<T = any> extends IChangeTarget<T>, IDisposable {}

/** The source of truth for an observable */
type Source<T = any> = (() => T) | T

let nextId = 1

/** An observable value */
export class Observable<T = any> implements IObserver<T> {
  ['_observers']?: Array<IChangeTarget<T>> = undefined
  protected _get?: () => T
  protected _value?: T

  readonly id: number

  constructor(source: Source<T>) {
    this.id = nextId++
    this._rebind(source)
  }

  /** Try coercing the given value into an Observable instance. */
  static from<T = any>(value: any): Observable<T> | undefined {
    return value instanceof Observable ? value : value ? value[$O] : undefined
  }

  /** The current value. */
  get(): T {
    return this._get ? this._get() : this._value!
  }

  /**
   * When the source of truth is a getter, this method only notifies observers.
   * Otherwise, this method also updates the current value.
   */
  set(newValue: T, force?: boolean) {
    let oldValue = this.get()
    if (force || newValue !== oldValue) {
      if (!this._get) {
        this._value = newValue
      }
      commit(this, oldValue, newValue)
    }
  }

  /** Listen for changes. */
  tap(onUpdate: (value: T) => void) {
    type ITapper = IObserver<T> & {
      onUpdate: typeof onUpdate
    }
    let observer: ITapper = {
      onUpdate,
      dispose: () => this._removeObserver(observer),
      _onChange: change =>
        change.prop === null && onUpdate(change.newValue as any),
    }
    this._addObserver(observer)
    return observer
  }

  /** Remove all observers and perform subclass cleanup (if necessary) */
  dispose() {
    if (this._observers) {
      this._observers = undefined
      this.deactivate()
    }
  }

  /** @internal Notify observers of a change */
  ['_onChange'](change: Change<T>) {
    let observers = this._observers
    if (getDebug(this)) {
      // tslint:disable-next-line
      console.debug(
        '%s(%s) changed: (observers: %O, change: %O)',
        this.constructor.name,
        this.id,
        observers ? observers.length : 0,
        shallowCopy(change) // TODO: Deep copy instead
      )
    }
    if (observers) {
      for (let observer of observers.slice(0)) {
        observer._onChange(change)
      }
    }
  }

  /** @internal Returns false when the given value is outdated */
  ['_isCurrent'](value: any) {
    return value === this.get()
  }

  /** @internal Swap out the source of truth for this observable. */
  ['_rebind'](source: Source<T>) {
    this._get = isFunction(source) ? source : undefined
    this._value = this._get ? undefined : (source as any)
  }

  /** @internal */
  ['_addObserver'](observer: IChangeTarget<T>, prepend?: boolean) {
    if (this._observers) {
      this._observers[prepend ? 'unshift' : 'push'](observer)
    } else {
      this.activate()
      this._observers = [observer]
    }
    return observer
  }

  /** @internal */
  ['_removeObserver'](observer: IChangeTarget<T>) {
    let observers = this._observers
    if (observers) {
      let index = observers.indexOf(observer)
      if (index < 0) return
      if (observers.length > 1) {
        observers.splice(index, 1)
      } else {
        this._observers = undefined
        this.deactivate()
      }
    }
  }

  /** Called when an Observable goes from 0 observers to 1+ */
  protected activate() {}

  /** Called when an Observable goes from 1+ observers to 0 */
  protected deactivate() {}
}

/** Returns true if the given value has been passed to `o()` and is not stale. */
export function isObservable(value: any): value is object {
  if (value) {
    let observable: Observable = value[$O]
    if (observable) {
      return observable._isCurrent(value)
    }
  }
  return false
}

/** Throws an error when the given object is an old revision */
export function getObservable<T = any>(value: object) {
  let observable: Observable<T> = value[$O]
  if (observable) {
    if (!observable._isCurrent(value)) {
      throw Error('Outdated values cannot be observed')
    }
    return observable
  }
}

/** @internal */
export function getObservers(value: object) {
  let observable = Observable.from(value)
  if (observable) {
    return observable._observers
  }
  throw Error('Not observable')
}
