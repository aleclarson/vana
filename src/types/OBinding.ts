import { IDisposable } from '../common'
import { Observable } from './Observable'

export type Driver<T = any> = (
  next: (value: T) => void
) => IDisposable | (() => void) | void

/**
 * An observable binding to some other stream-like mechanism.
 *
 * The binding is lazy, which means no value exists until observed.
 *
 * The current value is always `undefined` when no observers exist.
 */
export class OBinding<T = any> extends Observable<T> {
  private _binding?: IDisposable

  constructor(public driver: Driver<T>) {
    super(undefined as any)
  }

  protected activate() {
    const result = this.driver(this.set.bind(this))
    if (result) {
      this._binding = typeof result == 'function' ? { dispose: result } : result
    }
  }

  protected deactivate() {
    this._value = undefined
    if (this._binding) {
      this._binding.dispose()
      this._binding = undefined
    }
  }
}
