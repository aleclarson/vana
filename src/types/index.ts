import { Tapped } from '../funcs/tap'
import { Observable } from './Observable'

/** Useful for `subscribe` functions built on `vana` */
export type Observed<T> = T extends Observable<infer U> ? U : Tapped<T>

export { Observable, IObserver } from './Observable'
export { OBinding, Driver } from './OBinding'
export { OPath } from './OPath'
export { OPromise } from './OPromise'
export { OProps } from './OProps'
export { OTuple, ObservedTuple } from './OTuple'
export { Change, IChangeTarget } from './Change'
