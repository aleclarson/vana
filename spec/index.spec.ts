// tslint:disable:no-construct
import {
  freeze,
  isObservable,
  o,
  Observable,
  revise,
  setAutoFreeze,
} from '../src'
import { OBinding } from '../src/types/OBinding'
import { getObservable } from '../src/types/Observable'
import { OPromise } from '../src/types/OPromise'

const { isFrozen } = Object

describe('o()', () => {
  beforeAll(() => {
    setAutoFreeze(true)
  })
  afterAll(() => {
    setAutoFreeze(false)
  })

  it('recursively affects all objects inside the root object', () => {
    let obj = {}
    let arr = [obj]
    let base = o({ arr })

    expect(base.arr).toBe(arr)
    expect(isFrozen(arr)).toBeTruthy()
    expect(isObservable(arr)).toBeTruthy()

    expect(base.arr[0]).toBe(obj)
    expect(isFrozen(obj)).toBeTruthy()
    expect(isObservable(obj)).toBeTruthy()
  })

  describe('when a promise is passed', () => {
    it('makes the promise observable', () => {
      let p = o(Promise.resolve())
      expect(isObservable(p)).toBeTruthy()
      expect(getObservable(p)).toBeInstanceOf(OPromise)
    })
  })

  describe('when a driver function is passed', () => {
    it('returns an OBinding', () => {
      let ob = o<number>(next => {})
      expect(ob).toBeInstanceOf(OBinding)
    })
  })

  const objects = {
    'null object': () => Object.create(null),
    'plain object': () => ({}),
    'plain array': () => [],
  }

  for (let name in objects) {
    let getValue = objects[name] as () => object

    describe(`when a ${name} is passed`, () => {
      it('may contain an observable object', () => {
        let inner = o({})
        let outer = getValue()
        outer[0] = inner
        expect(() => o(outer)).not.toThrow()
        expect(isObservable(outer)).toBeTruthy()
        expect(isObservable(inner)).toBeTruthy()
      })

      describe('when value is already observable', () => {
        it('returns an observable copy', () => {
          let base = o(getValue())
          let copy = o(base)
          expect(isObservable(base)).toBeTruthy()
          expect(isObservable(copy)).toBeTruthy()
          expect(copy).not.toBe(base)
          expect(copy).toEqual(base)
        })

        it('even works with old revisions', () => {
          let base = o(getValue())
          let copy1 = revise(base, d => {
            d[0] = true
          })
          expect(isObservable(base)).toBeFalsy()
          let copy2 = o(base)
          expect(copy2).not.toBe(copy1)
          expect(copy2).not.toBe(base)
          expect(copy2).toEqual(base)
        })

        it('deeply clones old revisions', () => {
          let base = getValue()
          base[0] = { a: 1 }
          base[1] = { a: 2 } // <= This one won't be revised.
          revise(o(base), d => {
            d[0].a = 2
          })
          let copy = o(base)
          expect(isObservable(copy[0])).toBeTruthy()
          expect(copy[0]).not.toBe(base[0])
          expect(copy[1]).toBe(base[1])
        })
      })

      describe('when value is not yet observable', () => {
        it('binds an Observable to it', () => {
          let base = o(getValue())
          expect(isObservable(base)).toBeTruthy()
        })

        it('makes the values readonly', () => {
          let base = o(getValue())
          expect(isFrozen(base)).toBeTruthy()
        })

        it('returns the given value', () => {
          let base = getValue()
          expect(o(base)).toBe(base)
        })

        it('throws when the given value is already frozen', () => {
          let base = freeze(getValue())
          expect(() => o(base)).toThrowErrorMatchingSnapshot()
        })
      })
    })
  }

  class Foo {}
  const primitives = {
    string: '',
    boolean: true,
    number: 0,
    symbol: Symbol(),
    'null value': null,
    'undefined value': undefined,

    // Currently, these types throw an error.
    date: new Date(),
    regexp: /^/,
    'class instance': new Foo(),
    'Map object': new Map(),
    'Set object': new Set(),
  }

  for (let name in primitives) {
    const value: unknown = primitives[name]

    describe(`when a ${name} is passed`, () => {
      if (value && typeof value === 'object') {
        // TODO: Could we do something better here?
        it('throws an error', () => {
          expect(() => {
            o(value)
          }).toThrowErrorMatchingSnapshot()
        })
      } else {
        it('returns an Observable', () => {
          expect(o(value)).toBeInstanceOf(Observable)
        })
      }
    })

    it(`skips ${name}s when nested in an object`, () => {
      let base = o({ value })
      expect(isObservable(base.value)).toBeFalsy()
    })
  }
})

describe('isObservable()', () => {
  it('returns true for objects that were just passed to `o()`', () => {
    expect(isObservable(o({}))).toBeTruthy()
  })

  it('returns false for objects never passed to `o()`', () => {
    expect(isObservable({})).toBeFalsy()
  })

  it('returns false for objects that were revised since being passed to `o()`', () => {
    let base = o({ a: 0 })
    revise(base, { a: 1 })
    expect(isObservable(base)).toBeFalsy()
  })
})

describe('new Observable()', () => {
  describe('when a getter is passed', () => {
    it('dictate its value on-demand', () => {
      let val = 1
      let get = jest.fn(() => val)
      let o = new Observable(get)
      expect(o.get()).toBe(1)
      val = 2
      expect(o.get()).toBe(2)
      expect(get).toBeCalledTimes(2)
    })
  })

  describe('set() method', () => {
    it('updates the current value', () => {
      let o = new Observable<number>(1)

      let spy = jest.fn()
      o.tap(spy)

      o.set(2)
      expect(o.get()).toBe(2)
      expect(spy).toBeCalledWith(2)
    })

    it('strictly compares the new/old values before notifying', () => {
      let o = new Observable<number>(1)

      let spy = jest.fn()
      o.tap(spy)

      o.set(1)
      expect(spy).not.toBeCalled()
    })

    describe('when the "source" is a getter', () => {
      it('notifies observers', () => {
        let o = new Observable<number>(() => 1)

        let spy = jest.fn()
        o.tap(spy)

        o.set(2)
        expect(spy).toBeCalledWith(2)
      })

      it('never overrides the getter', () => {
        let o = new Observable<number>(() => 1)

        o.set(2)
        expect(o.get()).toBe(1)
      })
    })
  })
})
