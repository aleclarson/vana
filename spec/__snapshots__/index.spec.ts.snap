// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`o() when a Map object is passed throws an error 1`] = `"This object has no observable type"`;

exports[`o() when a Set object is passed throws an error 1`] = `"This object has no observable type"`;

exports[`o() when a class instance is passed throws an error 1`] = `"This object has no observable type"`;

exports[`o() when a date is passed throws an error 1`] = `"This object has no observable type"`;

exports[`o() when a null object is passed when value is not yet observable throws when the given value is already frozen 1`] = `"Frozen objects cannot become observable"`;

exports[`o() when a plain array is passed when value is not yet observable throws when the given value is already frozen 1`] = `"Frozen objects cannot become observable"`;

exports[`o() when a plain object is passed when value is not yet observable throws when the given value is already frozen 1`] = `"Frozen objects cannot become observable"`;

exports[`o() when a regexp is passed throws an error 1`] = `"This object has no observable type"`;
